data-analytics/iot-nirvana/client/src/main/java/com/google/cloud/demo/iot/nirvana/client/IotDevice.java

```diff
@@ -0,0 +1,201 @@
+  protected void generateDeviceId() {
+    SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
+    deviceId = "dev-" + sdf.format(new Date()) + "-" + UUID.randomUUID().toString();
+  }
+
+  /** Create the canonical IoT device registry path on GCP. */
+  protected String generateRegistryPath() throws ClientException {
+    return String.format(
+        "projects/%s/locations/%s/registries/%s",
+        getOptions().getGcpProjectId(),
+        getOptions().getGcpRegion(),
+        getOptions().getRegistryName());
+  }
+
+  /** Read public key and generate device credentials. */
+  private DeviceCredential loadCredentials() throws ClientException {
+    try {
+      PublicKeyCredential publicKeyCredential = new PublicKeyCredential();
+      String key =
+          Files.toString(new File(getOptions().getRsaCertificateFilePath()), Charsets.UTF_8);
+      publicKeyCredential.setKey(key);
+      publicKeyCredential.setFormat(KEY_FORMAT_RSA_X509_PEM);
+      DeviceCredential deviceCredential = new DeviceCredential();
+      deviceCredential.setPublicKey(publicKeyCredential);
+      return deviceCredential;
+    } catch (IOException e) {
+      String message =
+          String.format("Failed to load device credentials. Cause : %s", e.getMessage());
+      logError(message);
+      throw new ClientException(message, e);
+    }
+  }
+
+  /** Generate the Cloud IoT Core service. */
+  private CloudIot createCloudIotService() throws ClientException {
+    try {
+      logDebug("Generating Cloud IoT Core client...");
+      GoogleCredential credential =
+          GoogleCredential.getApplicationDefault().createScoped(CloudIotScopes.all());
+      JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
+      HttpRequestInitializer init = new RetryHttpInitializerWrapper(credential);
+      CloudIot service =
+          new CloudIot.Builder(GoogleNetHttpTransport.newTrustedTransport(), jsonFactory, init)
+              .setApplicationName(APP_NAME_API)
+              .build();
+      logDebug("Cloud IoT Core client generated.");
+      return service;
+    } catch (IOException e) {
+      String message =
+          String.format("Failed to create Cloud IoT Core service. Cause : %s", e.getMessage());
+      logError(message);
+      throw new ClientException(message, e);
+    } catch (GeneralSecurityException e) {
+      String message =
+          String.format("Failed to create Cloud IoT Core service. Cause : %s", e.getMessage());
+      logError(message);
+      throw new ClientException(message, e);
+    }
+  }
+
+  /** Log a debug message. */
+  private void logDebug(String message) throws ClientException {
+    LogUtils.writeLog(
+        logging,
+        message,
+        Severity.INFO,
+        deviceId,
+        LogUtils.CLOUD_LOGGING_DEVICE_SIMULATOR,
+        getOptions().getGcpProjectId());
+  }
+
+  /** Log an error message. */
+  private void logError(String message) throws ClientException {
```

Comment:
Why not move this as a static method to LogUtils so you dont repeat the same code in multiple classes?

Reply:
Done in commit aa68e7f20fc4feb6103d063226e7caeb3f4cc266